File: backend/app/schemas/admin_users.py
"""Schemas for admin user management"""

from datetime import datetime
from typing import Optional, List, Any
from pydantic import BaseModel, Field
from enum import Enum

from app.models.user import UserRole, UserTier, SubscriptionTier


# ============ Request Schemas ============

class UserRoleChangeRequest(BaseModel):
    """Request to change a user's role"""
    role: UserRole
    reason: Optional[str] = None


class BanUserRequest(BaseModel):
    """Request to ban a user"""
    reason: str = Field(..., min_length=10, description="Reason for banning (required)")


class SuspendUserRequest(BaseModel):
    """Request to suspend a user temporarily"""
    reason: str = Field(..., min_length=10, description="Reason for suspension")
    duration_hours: int = Field(..., ge=1, le=8760, description="Suspension duration in hours (max 1 year)")


class KarmaAdjustRequest(BaseModel):
    """Request to adjust a user's karma"""
    amount: int = Field(..., description="Amount to add (positive) or subtract (negative)")
    reason: str = Field(..., min_length=5, description="Reason for karma adjustment")


class TierOverrideRequest(BaseModel):
    """Request to override a user's tier"""
    tier: UserTier
    reason: Optional[str] = None


class UserSearchParams(BaseModel):
    """Parameters for searching/filtering users"""
    query: Optional[str] = None  # Search in email, full_name
    role: Optional[UserRole] = None
    tier: Optional[UserTier] = None
    subscription: Optional[SubscriptionTier] = None
    is_banned: Optional[bool] = None
    is_suspended: Optional[bool] = None
    is_verified: Optional[bool] = None
    sort_by: str = "created_at"
    sort_order: str = "desc"
    page: int = 1
    page_size: int = 20


# ============ Response Schemas ============

class UserListItem(BaseModel):
    """User item in list view"""
    id: int
    email: str
    full_name: Optional[str]
    avatar_url: Optional[str]
    role: UserRole
    user_tier: UserTier
    subscription_tier: SubscriptionTier
    karma_points: int
    is_active: bool
    is_verified: bool
    is_banned: bool
    is_suspended: bool
    suspended_until: Optional[datetime]
    total_reviews_given: int
    total_reviews_received: int
    created_at: datetime
    last_login: Optional[datetime]

    class Config:
        from_attributes = True


class UserListResponse(BaseModel):
    """Paginated user list response"""
    users: List[UserListItem]
    total: int
    page: int
    page_size: int
    total_pages: int


class UserDetailResponse(BaseModel):
    """Detailed user information for admin view"""
    id: int
    email: str
    full_name: Optional[str]
    avatar_url: Optional[str]
    bio: Optional[str]
    title: Optional[str]
    role: UserRole
    user_tier: UserTier
    subscription_tier: SubscriptionTier
    subscription_status: Optional[str]

    # Stats
    karma_points: int
    xp_points: int
    reputation_score: int
    total_reviews_given: int
    total_reviews_received: int
    avg_rating: Optional[float]
    acceptance_rate: Optional[float]
    current_streak: int
    longest_streak: int

    # Challenges
    challenges_won: int
    challenges_lost: int
    challenges_drawn: int

    # Status
    is_active: bool
    is_verified: bool
    is_banned: bool
    banned_at: Optional[datetime]
    ban_reason: Optional[str]
    is_suspended: bool
    suspended_until: Optional[datetime]
    suspended_at: Optional[datetime]
    suspension_reason: Optional[str]

    # Timestamps
    created_at: datetime
    updated_at: datetime
    last_login: Optional[datetime]

    class Config:
        from_attributes = True


class ModerationActionResponse(BaseModel):
    """Response for moderation actions"""
    success: bool
    message: str
    user_id: int
    action: str


class AdminStatsResponse(BaseModel):
    """Admin dashboard statistics"""
    total_users: int
    new_users_this_week: int
    active_users_today: int
    total_reviewers: int
    total_admins: int
    banned_users: int
    suspended_users: int
    pending_applications: int
    approved_this_month: int
    rejected_this_month: int
    active_challenges: int
    total_reviews: int
    avg_review_time_days: float


# ============ Audit Log Schemas ============

class AuditLogEntry(BaseModel):
    """Single audit log entry"""
    id: int
    admin_id: int
    admin_email: Optional[str]
    admin_name: Optional[str]
    action: str
    target_user_id: Optional[int]
    target_user_email: Optional[str]
    target_entity_type: Optional[str]
    target_entity_id: Optional[int]
    details: Optional[dict]
    ip_address: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True


class AuditLogResponse(BaseModel):
    """Paginated audit log response"""
    entries: List[AuditLogEntry]
    total: int
    page: int
    page_size: int


class BannedUserListResponse(BaseModel):
    """List of banned users"""
    users: List[UserListItem]
    total: int


class SuspendedUserListResponse(BaseModel):
    """List of suspended users"""
    users: List[UserListItem]
    total: int

File: frontend/lib/api/admin-users.ts
/**
 * Admin Users API Client
 * Handles user management operations for admins
 */

import apiClient from './client';

// ============ Types ============

export type UserRole = 'creator' | 'reviewer' | 'admin';
export type UserTier = 'novice' | 'contributor' | 'skilled' | 'trusted_advisor' | 'expert' | 'master';
export type SubscriptionTier = 'free' | 'pro';

export interface UserListItem {
  id: number;
  email: string;
  full_name: string | null;
  avatar_url: string | null;
  role: UserRole;
  user_tier: UserTier;
  subscription_tier: SubscriptionTier;
  karma_points: number;
  is_active: boolean;
  is_verified: boolean;
  is_banned: boolean;
  is_suspended: boolean;
  suspended_until: string | null;
  total_reviews_given: number;
  total_reviews_received: number;
  created_at: string;
  last_login: string | null;
}

export interface UserListResponse {
  users: UserListItem[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface UserDetail extends UserListItem {
  bio: string | null;
  title: string | null;
  subscription_status: string | null;
  xp_points: number;
  reputation_score: number;
  avg_rating: number | null;
  acceptance_rate: number | null;
  current_streak: number;
  longest_streak: number;
  challenges_won: number;
  challenges_lost: number;
  challenges_drawn: number;
  banned_at: string | null;
  ban_reason: string | null;
  suspended_at: string | null;
  suspension_reason: string | null;
  updated_at: string;
}

export interface AdminStats {
  total_users: number;
  new_users_this_week: number;
  active_users_today: number;
  total_reviewers: number;
  total_admins: number;
  banned_users: number;
  suspended_users: number;
  pending_applications: number;
  approved_this_month: number;
  rejected_this_month: number;
  active_challenges: number;
  total_reviews: number;
  avg_review_time_days: number;
}

export interface AuditLogEntry {
  id: number;
  admin_id: number;
  admin_email: string | null;
  admin_name: string | null;
  action: string;
  target_user_id: number | null;
  target_user_email: string | null;
  target_entity_type: string | null;
  target_entity_id: number | null;
  details: Record<string, unknown> | null;
  ip_address: string | null;
  created_at: string;
}

export interface AuditLogResponse {
  entries: AuditLogEntry[];
  total: number;
  page: number;
  page_size: number;
}

export interface ModerationActionResponse {
  success: boolean;
  message: string;
  user_id: number;
  action: string;
}

export interface UserSearchParams {
  query?: string;
  role?: UserRole;
  tier?: UserTier;
  is_banned?: boolean;
  is_suspended?: boolean;
  is_verified?: boolean;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
  page?: number;
  page_size?: number;
}

// ============ API Functions ============

export const adminUsersApi = {
  /**
   * List users with filtering and pagination
   */
  listUsers: async (params: UserSearchParams = {}): Promise<UserListResponse> => {
    const searchParams = new URLSearchParams();

    if (params.query) searchParams.set('query', params.query);
    if (params.role) searchParams.set('role', params.role);
    if (params.tier) searchParams.set('tier', params.tier);
    if (params.is_banned !== undefined) searchParams.set('is_banned', String(params.is_banned));
    if (params.is_suspended !== undefined) searchParams.set('is_suspended', String(params.is_suspended));
    if (params.is_verified !== undefined) searchParams.set('is_verified', String(params.is_verified));
    if (params.sort_by) searchParams.set('sort_by', params.sort_by);
    if (params.sort_order) searchParams.set('sort_order', params.sort_order);
    if (params.page) searchParams.set('page', String(params.page));
    if (params.page_size) searchParams.set('page_size', String(params.page_size));

    return apiClient.get(`/admin/users?${searchParams.toString()}`);
  },

  /**
   * Get admin dashboard statistics
   */
  getStats: async (): Promise<AdminStats> => {
    return apiClient.get('/admin/users/stats');
  },

  /**
   * Get user detail
   */
  getUser: async (userId: number): Promise<UserDetail> => {
    return apiClient.get(`/admin/users/${userId}`);
  },

  /**
   * Change user role
   */
  changeRole: async (userId: number, role: UserRole, reason?: string): Promise<ModerationActionResponse> => {
    return apiClient.post(`/admin/users/${userId}/role`, { role, reason });
  },

  /**
   * Ban a user
   */
  banUser: async (userId: number, reason: string): Promise<ModerationActionResponse> => {
    return apiClient.post(`/admin/users/${userId}/ban`, { reason });
  },

  /**
   * Unban a user
   */
  unbanUser: async (userId: number): Promise<ModerationActionResponse> => {
    return apiClient.post(`/admin/users/${userId}/unban`);
  },

  /**
   * Suspend a user temporarily
   */
  suspendUser: async (userId: number, reason: string, durationHours: number): Promise<ModerationActionResponse> => {
    return apiClient.post(`/admin/users/${userId}/suspend`, {
      reason,
      duration_hours: durationHours,
    });
  },

  /**
   * Remove suspension from a user
   */
  unsuspendUser: async (userId: number): Promise<ModerationActionResponse> => {
    return apiClient.post(`/admin/users/${userId}/unsuspend`);
  },

  /**
   * Verify a user's email
   */
  verifyUser: async (userId: number): Promise<ModerationActionResponse> => {
    return apiClient.post(`/admin/users/${userId}/verify`);
  },

  /**
   * Adjust user's karma
   */
  adjustKarma: async (userId: number, amount: number, reason: string): Promise<ModerationActionResponse> => {
    return apiClient.post(`/admin/users/${userId}/karma`, { amount, reason });
  },

  /**
   * Override user's tier
   */
  overrideTier: async (userId: number, tier: UserTier, reason?: string): Promise<ModerationActionResponse> => {
    return apiClient.post(`/admin/users/${userId}/tier`, { tier, reason });
  },

  /**
   * Get banned users list
   */
  getBannedUsers: async (): Promise<{ users: UserListItem[]; total: number }> => {
    return apiClient.get('/admin/users/moderation/banned');
  },

  /**
   * Get suspended users list
   */
  getSuspendedUsers: async (): Promise<{ users: UserListItem[]; total: number }> => {
    return apiClient.get('/admin/users/moderation/suspended');
  },

  /**
   * Get audit log
   */
  getAuditLog: async (params: {
    admin_id?: number;
    target_user_id?: number;
    page?: number;
    page_size?: number;
  } = {}): Promise<AuditLogResponse> => {
    const searchParams = new URLSearchParams();
    if (params.admin_id) searchParams.set('admin_id', String(params.admin_id));
    if (params.target_user_id) searchParams.set('target_user_id', String(params.target_user_id));
    if (params.page) searchParams.set('page', String(params.page));
    if (params.page_size) searchParams.set('page_size', String(params.page_size));

    return apiClient.get(`/admin/users/moderation/audit?${searchParams.toString()}`);
  },
};

export default adminUsersApi;
